### ISSUE TITLE
"java.lang.IllegalArgumentException: Bad escape" when path has non-ASCII characters

### ISSUE THREAD

**Issue by alexchentt** [redhat-developer/vscode-java#3735]
### EDIT: CORRECTION
**TLDR is that on version 1.33.0, when the project path contains non-ASCII characters, diagnostics would throw the exception `java.lang.IllegalArgumentException: Bad escape`. Moving the project to a path that contains only ASCII characters would prevent this.**

### Original Post
[provide a description of the issue]
(Noobie here, I'm sorry if I am not supposed to ask about this here)

Fresh install of openjdk@21, vscode (universal binary) and Extension Pack for Java (that includes redhat-developer/vscode-java 1.33.0) on arm64 macOS 15.0 will cause internal errors when loading in any java projects and the plugin will then provide no syntax checking. **The plugin on 1.32.0 functions as expected.**

Most errors are of the same type, here is an example:
```
!ENTRY org.eclipse.jdt.ls.core 1 0 2024-08-06 17:03:04.096
!MESSAGE >> java/validateDocument

!ENTRY org.eclipse.jdt.ls.core 1 0 2024-08-06 17:03:04.711
!MESSAGE begin problem for /Sample.java

!ENTRY org.eclipse.core.jobs 4 2 2024-08-06 17:03:04.712
!MESSAGE An internal error occurred during: "Publish Diagnostics".
!STACK 0
java.lang.IllegalArgumentException: Bad escape
	at java.base/sun.nio.fs.UnixUriUtils.fromUri(Unknown Source)
	at java.base/sun.nio.fs.UnixFileSystemProvider.getPath(Unknown Source)
	at java.base/java.nio.file.Path.of(Unknown Source)
	at java.base/java.nio.file.Paths.get(Unknown Source)
	at org.eclipse.jdt.ls.core.internal.JDTUtils.isExcludedFile(JDTUtils.java:1843)
	at org.eclipse.jdt.ls.core.internal.handlers.BaseDiagnosticsHandler.matchesDiagnosticFilter(BaseDiagnosticsHandler.java:154)
	at org.eclipse.jdt.ls.core.internal.handlers.WorkspaceDiagnosticsHandler.publishDiagnostics(WorkspaceDiagnosticsHandler.java:324)
	at org.eclipse.jdt.ls.core.internal.handlers.WorkspaceDiagnosticsHandler.publishDiagnostics(WorkspaceDiagnosticsHandler.java:254)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer$2.run(JDTLanguageServer.java:334)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

##### Environment
- Operating System: macOS 15.0 Beta (24A5298h) arm64
- JDK version: OpenJDK 21.0.4
- Visual Studio Code version: Version: 1.92.0 (Universal)
- Java extension version: 1.33.0 (1.32.0 works fine)

##### Steps To Reproduce
1. Install macOS 15.0 Beta.
2. Fresh install OpenJDK `brew install openjdk@21`
3. Symlink and path setting of openjdk
```
sudo ln -sfn /opt/homebrew/opt/openjdk@21/libexec/openjdk.jdk /Library/Java/JavaVirtualMachines/openjdk-21.jdk)
echo 'export PATH="/opt/homebrew/opt/openjdk@21/bin:$PATH"' >> ~/.profile
```
4. Fresh install vscode
5. Install Extension Pack for Java in vscode
6. ~~Load any projects or create a new one~~ **Load any projects WITH PATH containing non-ASCII characters***
7. Let the extension load
8. Run "Open Java Language Server Logfile" and see the error.

[Please attach a sample project reproducing the error]
~~Any project will do.~~ Projects located in paths with non-ASCII characters

[Please attach logs](https://github.com/redhat-developer/vscode-java/wiki/Troubleshooting#enable-logging)
[log.log](https://github.com/user-attachments/files/16508155/log.log)

**Comment by rgrunber** [comment]
Are you using `java.diagnostic.filter` option ? What is it set to ?

**Comment by alexchentt** [comment]
@rgrunber that non ASCII character bug is the culprit, thank you!

My new system is German and turns out the OneDrive directory is not ~/Library/CloudStorage/OneDrive-Personal but OneDrive-Persönlich. That ö is causing all the trouble. After moving the folder out of OneDrive, plugin version 1.33.0 worked absolutely flawlessly. Thank you so much.

**Comment by rgrunber** [comment]
We added a [new feature](https://github.com/redhat-developer/vscode-java/issues/2150) in 1.33.0 that requires checking each file (most likely when diagnostics need to be update). This happens in places like : https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseDiagnosticsHandler.java#L149-L155 . This ends up calling https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java#L1837-L1846 .

We can definitely reduce the likelihood of this for people not using the filter feature, since we can check if the list is empty before going through all that conversion.

**Comment by snjeza** [comment]
We can try to  use
```
path = Paths.get(uri.toURL().getPath());
```
instead of
```
path = Paths.get(uri);
```
at https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/b7998dc6e5c42fe7b83c69d2eab3b150d75ff025/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java#L1843

**Comment by rgrunber** [comment]
I can add that to the existing PR after testing that it works.

**Comment by CT702-7** [comment]
Same issue happens when chinese characters appear in the address of the project...
